dbos_transact/dbos.py

```diff
@@ -273,26 +425,45 @@ def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
 
                     return self._execute_workflow(status, func, *args, **kwargs)
 
-        wrapped_func = cast(Workflow[P, R], wrapper)
+        wrapped_func = cast(F, wrapper)
         return wrapped_func
 
-    def register_wf_function(self, name: str, wrapped_func: Workflow[P, R]) -> None:
+    def _register_wf_function(self, name: str, wrapped_func: F) -> None:
         self.workflow_info_map[name] = wrapped_func
 
-    def workflow_decorator(self, func: Workflow[P, R]) -> Workflow[P, R]:
-        wrapped_func = self.workflow_wrapper(func)
-        self.register_wf_function(func.__qualname__, wrapped_func)
+    def _workflow_decorator(self, func: F) -> F:
+        wrapped_func = self._workflow_wrapper(func)
+        self._register_wf_function(func.__qualname__, wrapped_func)
         return wrapped_func
 
-    def workflow(self) -> Callable[[Workflow[P, R]], Workflow[P, R]]:
-        return self.workflow_decorator
+    def workflow(self) -> Callable[[F], F]:
+        return self._workflow_decorator
+
+    # There seems to be a bug / limitation in mypy regarding class / instance arguments
+    #   Not allowing anything besides Workflow[P,R] leads to false errors in mypy
+    #   Providing these overloads allows too much in mypy, though most errors are still caught
+    #   The pytest / runtime error is clear.
+    @overload
+    def start_workflow(
+        self, func: WorkflowC[P, R], cls: type, *args: P.args, **kwargs: P.kwargs
+    ) -> WorkflowHandle[R]: ...
+
+    @overload
+    def start_workflow(
+        self, func: WorkflowI[P, R], inst: object, *args: P.args, **kwargs: P.kwargs
```

Comment:
For "instances", we need to record `config_name` in the `dbos.workflow_status` table for workflow recovery.
And then the recovery code would need to correctly reconstruct the class instances.

Maybe in a separate PR?

Reply:
The recording part is done and tested, comment below about recovery is open